Sming/Arch/Esp32/Core/HardwarePWM.cpp

```diff
@@ -0,0 +1,345 @@
+			ledc_bind_channel_timer(pinToGroup(i), pinToChannel(i), pinToTimer(i));
+		}
+		maxduty=maxDuty(DEFAULT_RESOLUTION);
+		const int initial_period=DEFAULT_PERIOD;
+	}
+}
+
+HardwarePWM::~HardwarePWM()
+{
+	// There is no function in the SDK to stop PWM output, yet.
+}
+
+/* Function Name: getChannel
+ * Description: This function is used to get channel number for given pin
+ * Parameters: pin - Esp8266 pin number
+ */
+uint8_t HardwarePWM::getChannel(uint8_t pin)
+{
+	for(uint8_t i = 0; i < channel_count; i++) {
+		if(channels[i] == pin) {
+			//debugf("getChannel %d is %d", pin, i);
+			return i;
+		}
+	}
+	return -1;
+}
+
+/* Function Name: getDutyChan
+ * Description: This function is used to get the duty cycle number for a given channel
+ * Parameters: chan -Esp8266 channel number
+ */
+uint32_t HardwarePWM::getDutyChan(uint8_t chan)
+{
+	if(chan == PWM_BAD_CHANNEL) {
+		return 0;
+	} else {
+		return ledc_get_duty(pinToGroup(chan),pinToChannel(chan));		
+	}
+		// esp32 defines the frequency / period per timer,  
+}
+
+/* Function Name: setDutyChan
+ * Description: This function is used to set the pwm duty cycle for a given channel. If parameter 'update' is false
+ *              then you have to call update() later to update duties.
+ * Parameters: chan - channel number
+ *             duty - duty cycle value
+ *             update - update PWM output
+ */
+bool HardwarePWM::setDutyChan(uint8_t chan, uint32_t duty, bool update)
+{
+	if(chan == PWM_BAD_CHANNEL) {
+		return false;
+	} else if(duty <= maxduty) {
+		ESP_ERROR_CHECK(ledc_set_duty(pinToGroup(chan), pinToChannel(chan), duty));
+		if(update) {
+			ESP_ERROR_CHECK(ledc_update_duty(pinToGroup(chan), pinToChannel(chan)));
+			//update();
+		}
+		return true;
+	} else {
+		debugf("Duty cycle value too high for current period.");
+		return false;
+	}
+}
+
+/* Function Name: getPeriod
+ * Description: This function is used to get Period of PWM.
+ *				Period / frequency will remain same for all pins.
+ *
+ */
+uint32_t HardwarePWM::getPeriod()
+{
+	// sming does not know how to handle different frequencies for channels, this will require an extended interface
+	// for now, just report the period for group 0 channel 0 
+	return frequencyToPeriod(ledc_get_freq(static_cast<ledc_mode_t>(0),static_cast<ledc_timer_t>(0)));
+}
+
+/* Function Name: setPeriod
+ * Description: This function is used to set Period of PWM.
+ *				Period / frequency will remain same for all pins.
+ */
+void HardwarePWM::setPeriod(uint32_t period)
+{
+	// setting the frequency globally, will add per timer functions later
+	// also, this can be done smarter 
+	for(uint8_t i = 0; i < channel_count; i++) {
+		ESP_ERROR_CHECK(ledc_set_freq(pinToGroup(i), pinToTimer(i), periodToFrequency(period)));
+	}
+	//sledc_update_duty();
+	update();
+}
+
+/* Function Name: update
+ * Description: This function is used to actually update the PWM.
+ */
+void HardwarePWM::update()
+{
+	//ledc_update_duty();	
```

Comment:
I believe the purpose of this method is so that multiple channels can be reconfigured, then updated all together. For some applications this may be important to ensure they are synchronised. Unfortunately the IDF HAL doesn't support this.

My personal preference here would be to strip away all the HAL code and implement using low-level calls (such as `ledc_ll_ls_channel_update`). For now, probably just add a note to the code.